cmd/controller/processevent.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2018 OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	apis "github.com/openebs/node-disk-manager/pkg/apis/ndm/v1alpha1"
+	"github.com/openebs/node-disk-manager/pkg/udev"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func processDiskFromUdevEvent(c *Controller, msg map[string]string) error {
+	label := fmt.Sprintf("kubernetes.io/hostname=%v", c.hostname)
+	filter := metav1.ListOptions{LabelSelector: label}
+	listDR, err := c.clientset.NdmV1alpha1().Disks().List(filter)
+	if err != nil {
+		return err
+	}
+	if msg["ACTION"] == "add" {
+		processDetailsForAttach(c, msg, listDR)
+	} else if msg["ACTION"] == "remove" {
+		processDetailsForDetach(c, msg, listDR)
+	}
+	return nil
+}
+
+func processDetailsForAttach(c *Controller, msg map[string]string, listDR *apis.DiskList) {
+	device := udev.NewDeviceFromSysPath(msg["SYSPATH"])
+	uuid := udev.GetUid(device)
+	if uuid == msg["uuid"] {
+		if device.Devtype() == "disk" {
+			dr, err := collectDiskInfo(device)
+			if err == nil {
+				dr.ObjectMeta.Labels[NDMHostKey] = c.hostname
+				edr := getExistingResource(uuid, listDR)
+				if edr != nil {
+					glog.Info("disk object already exist in etcd : ", uuid)
+					updateDR(dr, edr, c)
+				} else {
+					createDR(dr, c)
+				}
+			} else {
+				glog.Info("Error while collecting disk details")
```

Comment:
Why is this Info?

Reply:
Can you please write the logic using `err != nil {log, logic & return}`,  ` if uuid != msg["uuid"] {logic & return}` ?
Please check the entire function & see the nested blocks. This is very difficult to read & maintain. 
We need to write logic that avoids nested blocks.

Reply:
Refactored code according to your comments.